docs/react/guides/ssr.md

```diff
@@ -151,371 +153,376 @@ function Posts() {
-  const { data } = useQuery({ queryKey: ['posts'], queryFn: getPosts })
-
-  // This query was not prefetched on the server and will not start
-  // fetching until on the client, both patterns are fine to mix
-  const { data: otherData } = useQuery({
-    queryKey: ['posts-2'],
-    queryFn: getPosts,
-  })
-
-  // ...
-}
+dehydrate(queryClient, {
+  shouldDehydrateQuery: (query) => {
+    // This will include all queries, including failed ones,
+    // but you can also implement your own logic by inspecting `query`
+    return true
+  },
+})
 ```
 
-As demonstrated, it's fine to prefetch some queries and let others fetch on the client. This means you can control what content server renders or not by adding or removing `prefetchQuery` for a specific query.
+## Serialization
 
-### Streaming, Suspense and server-side fetching
+When doing `return { props: { dehydratedState: dehydrate(queryClient) } }` in Next.js, or `return json({ dehydratedState: dehydrate(queryClient) })` in Remix, what happens is that the `dehydratedState` representation of the `queryClient` is serialized by the framework so it can be embedded into the markup and transported to the client.
 
-Right now, you always have to `await` the data in the Server Component. In the future, the goal is to be able to _start_ prefetching in a Server Component but not block rendering, instead streaming markup and data to the client incrementally as it gets available. This is currently lacking support in both React and Query.
+By default, these frameworks only supports returning things that are safely serializable/parsable, and therefor does not support `undefined`, `Date`, `Map`, `Set`, `BigInt`, `Infinity`, `NaN`, `-0`, regular expressions etc. This also means that you can not return any of these things from your queries. If returning these values is something you want, check out [superjson](https://github.com/blitz-js/superjson) or similar packages.
 
-Similarily, you _must_ currently prefetch the data in a Server Component if you want it to be server rendered. A `useQuery()` call even with the `suspense` option enabled will not fetch data on the server, only on the client. We hope to support this in the future, but exact details are still unknown.
+If you are using a custom SSR setup, you need to take care of this step yourself. Your first instinct might be to use `JSON.stringify(dehydratedState)`, but because this doesn't escape things like `<script>alert('Oh no..')</script>` by default, this can easily lead to **XSS-vulnerabilities** in your application. [superjson](https://github.com/blitz-js/superjson) also **does not** escape values and is unsafe to use by itself in a custom SSR setup (unless you add an extra step for escaping the output). Instead we recommend using a library like [Serialize JavaScript](https://github.com/yahoo/serialize-javascript) or [devalue](https://github.com/Rich-Harris/devalue) which are both safe against XSS injections out of the box.
```

Comment:
TIL 😱

Reply:
Not sure which part was the TIL, but I learned this the hard way. 😬 It was easily the most common React vulnerability back when custom SSR was more common (might still be).

Say you make a url-parameter part of the `queryKey` (like an id). Now anyone can craft a link that contains `?articleId=<script>BAD STUFF</script>` and spread it, and if that is not escaped and included in the markup for hydration, now that JS is executed in the visitors browser. Same with user generated data ofc unless it's validated/escaped in the backend.

Reply:
> Not sure which part was the TIL

everything really 😅